### ISSUE TITLE
[🐞] Uncaught (in promise) TypeError: Failed to resolve module specifier "X". Relative references must start with either "/", "./", or "../".

### ISSUE THREAD

**Issue by Nefcanto** [QwikDev/qwik#5281]
### Which component is affected?

Qwik Runtime

### Describe the bug

I'm creating a page using Qwik, and when I add client interaction to it using `useSignal` and `onClick$` functions, I get this error in my client in the production code:

> Uncaught (in promise) TypeError: Failed to resolve module specifier "BrandsParts". Relative references must start with either "/", "./", or "../".

The problem is that the `BrandsParts` is not part of this page. It's not related to this page at all.

Based on [this](https://stackoverflow.com/a/62282239/811785) answer and [this](https://stackoverflow.com/a/72458960/811785) answer, it seems that Qwik is trying to load one of our imports on the client-side, inside the browser:

    import { Layout as RunnableLayout } from 'BrandsParts'

This code works just fine in the development environment, and I have no clue why it breaks and behaves such when it's built for production.

### Reproduction

I can give you access to our private repo.

### System Info

```shell
## Production server
  System:
    OS: Linux 5.10 Debian GNU/Linux 11 (bullseye) 11 (bullseye)
    CPU: (4) x64 Intel(R) Xeon(R) CPU E5-2680 v4 @ 2.40GHz
    Memory: 84.48 MB / 9.73 GB
    Container: Yes
    Shell: 5.1.4 - /bin/bash
  Binaries:
    Node: 18.18.0 - /usr/local/bin/node
    Yarn: 1.22.19 - /usr/local/bin/yarn
    npm: 9.8.1 - /usr/local/bin/npm

## Development machine
  System:
    OS: Linux 6.2 Debian GNU/Linux 11 (bullseye) 11 (bullseye)
    CPU: (4) x64 Intel(R) Core(TM) i5-6500 CPU @ 3.20GHz
    Memory: 9.69 GB / 15.49 GB
    Container: Yes
    Shell: 5.1.4 - /bin/bash
  Binaries:
    Node: 18.18.0 - /usr/local/bin/node
    Yarn: 1.22.19 - /usr/local/bin/yarn
    npm: 10.1.0 - /usr/local/bin/npm
  Browsers:
    Chromium: 117.0.5938.132
  npmPackages:
    @builder.io/partytown: ^0.8.1 => 0.8.1 
    @builder.io/qwik: ^1.2.13 => 1.2.12 
    @builder.io/qwik-auth: ~0.1.3 => 0.1.3 
    @builder.io/qwik-city: ^1.2.13 => 1.2.12 
    undici: ^5.25.4 => 5.25.2 
    vite: ^4.4.11 => 4.4.9
```

**Comment by wmertens** [comment]
Is this from a library? Sounds like a bundling issue due to vite configuration

**Comment by Nefcanto** [comment]
No. In fact, I have no idea what is the source of this problem. Our `Vite` configuration is this:

```



    return {
        plugins: [
            qwikCity({
                trailingSlash: false
            }),
            qwikVite(),
            tsconfigPaths(),
            partytownVite({ dest: join(__dirname, "dist", "~partytown") }),
        ],
        preview: {
            headers: {
                'Cache-Control': 'public, max-age=600',
            },
        },
        server: {
            host: "0.0.0.0",
            hmr: {
                clientPort: 443,
            },
        },
        build: {
            rollupOptions: {
                external: [
                    ...barrels
                ]
            }
        },
        optimizeDeps: {
            include: ["@auth/core"]
        }
    }
```

**Comment by wmertens** [comment]
The ...barrels could be an issue

**Comment by Nefcanto** [comment]
We realized that this is related to the name of the `index.jsx` component being either `Index` or a more semantic name.

If we create a page for `/blog/[...post]` and name the component as `const Blog = component$` we won't get errors.

However, when we rename it to `const Index = component$` it throws that error.

What could be wrong?

**Comment by mhevery** [comment]
> I can give you access to our private repo.

We need a small stand-alone reproduction case. Could you please provide one.

**Comment by Nefcanto** [comment]
@mhevery  we tried a couple of times, but we couldn't make an MRE.

**Comment by mhevery** [comment]
You can see how without a reproducible case it would be hard for us to help.

**Comment by Nefcanto** [comment]
@mhevery, yes I understand that. And if you look at my issues we have tried to create an MRE many times. But this time, it would be an exception if you help us with a real project.

The point is, that we left Next.js completely and migrated to Qwik. We are a web design company. This means we need to make websites for many customers, which in turn means creating a meta-meta-framework. Qwik is THE framework, Qwik-City is the meta-framework, and our internal framework is the meta-meta-framework.

For example, you can see that all of these websites of our clients are built using your amazing framework:

[https://afsharmotors.com/](https://afsharmotors.com/)
[https://e-learningland.com/](https://e-learningland.com/)
[https://hypersanat.net/](https://hypersanat.net/)
[https://new.zisttakhmir.com](https://new.zisttakhmir.com/)
[https://behsoo.ir/](https://behsoo.ir/)
https://new.amlakmosallas.com/
[https://technonata.com/](https://technonata.com/)
[https://mobinzar.com/](https://mobinzar.com/)
[https://pydr.ir/](https://pydr.ir/)
[https://holism.ir/](https://holism.ir/)
```

I sent this list only to prove that we're investing heavily in Qwik as our only client framework. But we can't create everything from scratch for all customers. We need to be able to reuse a lot of things. And because Qwik did not have a good extensibility architecture built into it for cross-project reusability, we had to somehow do some creative works.

So, I think it can be a good point that we help each other in making Qwik a more dynamic framework that can support more scenarios.

You can understand our needs as a real-world company that deals with tons of requirements from real customers, and that can help Qwik become even better.

Yet if you do not have time for this, I completely understand and still thank you for this revolutionary hydration-less framework.

**Comment by wmertens** [comment]
@Nefcanto Can you try making a MRE by copying your private repo and removing parts until the problem stops occurring?

Right now we don't even know where in the qwik code the problem is...

**Comment by Nefcanto** [comment]
@wmertens that's a very good idea. Thank you for suggesting it. Yep, I'll do it. It might take a while because it's completely heuristical. Yet I'll come back with an MRE.

**Comment by brandonpittman** [comment]
Getting something similar when using Vanilla Extract Sprinkles at runtime.
<img width="879" alt="Screenshot 2023-10-26 at 3 49 21 PM" src="https://github.com/BuilderIO/qwik/assets/967145/fcf2c420-4bf6-42f3-9e84-feb38105c0f1">

**Comment by wmertens** [comment]
@brandonpittman that's very odd, it didn't bundle the sprinkles module? 
Might be useful to add vite-plugin-inspect to your config and see how that calling module gets transformed

**Comment by brandonpittman** [comment]
@wmertens 

While running the dev server, I saw this pop up in server console:

`[vite] ✨ new dependencies optimized: @vanilla-extract/sprinkles/createRuntimeSprinkles`

Connected, I suppose?

**Comment by wmertens** [comment]
So in prod, somehow vite doesn't find the dependency. I can't think of how that happens, seems like a vite bug

**Comment by brandonpittman** [comment]
Gotta be when it tries this line:

https://github.com/vanilla-extract-css/vanilla-extract/blob/1dbd56c8292af03e53aa3f4b288a40d88bc762cf/packages/sprinkles/src/index.ts#L370-L375

**Comment by wmertens** [comment]
@brandonpittman try using the vite plugin from styled-vanilla-extract?

The function you point at generates an import at build time, and then vite should still see it and process it. The inspect plugin should also show the generated code.

**Comment by brandonpittman** [comment]
> try using the vite plugin from styled-vanilla-extract?

Just did—same result.

**Comment by wmertens** [comment]
So the question is, if you disable the qwik vite plugin(s), does it still have a string reference to `@vanilla-extract/sprinkles/createRuntimeSprinkles` in the build? (The build won't work mind you)

**Comment by brandonpittman** [comment]
> if you disable the qwik vite plugin(s), does it still have a string reference to @vanilla-extract/sprinkles/createRuntimeSprinkles in the build? 

grepped for "createRuntime" and found nothing.

**Comment by wmertens** [comment]
ok, so now you need to use the vite-plugin-inspect to compare this build with one where the qwik plugins are enabled. You can copy the .vite-inspect directory to see them side-by-side.

It's still very strange to me that the qwik plugin can emit imports without vite seeing them, maybe vite thinks it's not js?

**Comment by brandonpittman** [comment]
This is in the Qwik build:

```
import { S as s, a as b, h as t, R as y } from "./q-4af319bc.js";
import "./q-4ed993c7.js";
import { createSprinkles as u } from "@vanilla-extract/sprinkles/createRuntimeSprinkles";
import { A as _ } from "./q-317c7fa7.js";
```

Is it weird to see a non-relative import in a build file?

**Comment by brandonpittman** [comment]
@wmertens @mhevery 

I made a private repo that shows the problem and gave you both access.

https://github.com/brandonpittman/qwik-ve-app-with-library


If you you `npm run preview` and click the "Tokens" link on the index route, you'll see the error in the console.

**Comment by wmertens** [comment]
- first of all, the import order was wrong, but that didn't fix things. (I pushed a commit)
- I'm not sure why qwik-vite-plugin transforms the atoms.css.ts file, it shouldn't touch it.
- with the inspect plugin you can see that first qwik-vite loads it, then vanilla-extract transforms it, adding the createSprinkles import, then qwik-vite transforms it, and then for some reason the import is ignored by vite.

**Comment by brandonpittman** [comment]
> first of all, the import order was wrong, but that didn't fix things. (I pushed a commit)

If I don't put `qwikCity()` before `qwikVite()`, I get an error about Missing Qwik City ENV data.

**Comment by brandonpittman** [comment]
Solved it! It's this:

```ts
rollupOptions: {
        // externalize deps that shouldn't be bundled into the library
        external: [
          /^node:.*/,
          ...excludeAll(dependencies),
          ...excludeAll(peerDependencies),
        ],
      },
    },
```

If I remove that, the Qwik City stuff all loads without trouble. I need it for the component library, but at least I know that removing that bit about externals will sort out the docs site I'm trying to build.

**Comment by brandonpittman** [comment]
What part of Qwik City generates the `service-worker.js`? That still doesn't seem to be getting generated.

**Comment by wmertens** [comment]
> Solved

You actually need this setting for libraries, but if you're building for local testing then you don't. So this setting should be adjusted to be only in lib mode

**Comment by wmertens** [comment]
@Nefcanto is this also the solution for you?

**Comment by brandonpittman** [comment]
> You actually need this setting for libraries, but if you're building for local testing then you don't. So this setting should be adjusted to be only in lib mode

I'll look into how to do that tomorrow. Thanks.

**Comment by brandonpittman** [comment]
@wmertens 

Is there a better solution than this?

```tsx
import { component$, useVisibleTask$ } from "@builder.io/qwik";
import { Link, routeLoader$ } from "@builder.io/qwik-city";

export const useLoader = routeLoader$((ev) => {
  return ev.env.get("NODE_ENV");
});

export default component$(() => {
  const data = useLoader();

  useVisibleTask$(() => {
    window.process = {
      env: {
        NODE_ENV: data.value,
      },
    } as any;
  });

  return <Link href="/tokens">Tokens: {data}</Link>;
});
```

**Comment by wmertens** [comment]
> Is there a better solution than this?

Sprinkles uses the env to know if it should build for prod or dev, so not runtime.

There should be a rollup option somewhere that allows injecting that value as a global?

**Comment by brandonpittman** [comment]
@wmertens 

```ts
    define: {
      "process.env.NODE_ENV": `'${process.env.NODE_ENV}'`,
    },
```

This works.

**Comment by wmertens** [comment]
> Also added a conditional based on Vite's mode which seems to fix the initial problem.

Could I persuade you to do a little PR for the lib starter? :pleading_face:

**Comment by brandonpittman** [comment]
> > Also added a conditional based on Vite's mode which seems to fix the initial problem.
> 
> Could I persuade you to do a little PR for the lib starter? 🥺

Sure. I had the same thought.

**Comment by Nefcanto** [comment]
> ```ts
> peerDependencies
> ```

Our team is creating a repo for this the way you mentioned. It would be ready these days. But I didn't understand what that `excludeAll` is.

**Comment by gioboa** [comment]
This issue is really old, is it still valid? if yes, can we a have a basic repo to reproduce this issue?

**Comment by wmertens** [comment]
Relative imports issues were solved a while ago, closing, please reopen if still an issue
